### PR TITLE
Add residency manager fuzzing.

### PR DESCRIPTION
**Pull request by bbernhar**
Adds a fuzzer for the residency LRUs by filling up the budget, then creating randomly sized buffers.

### CODE REVIEW

**Comment by bbernhar** [comment]
@bjjones Could you PTAL?

File: src/fuzzers/D3D12ResidencyManagerFuzzer.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    const DXGI_MEMORY_SEGMENT_GROUP bufferMemorySegment =
+        gResidencyManager->GetMemorySegmentGroup(allocationDesc.HeapType);
+
+    constexpr uint64_t kBufferMemorySize = GPGMM_MB_TO_BYTES(1);
+    const D3D12_RESOURCE_DESC bufferDesc = CreateBufferDesc(kBufferMemorySize);
+
+    // Keep allocating until we reach the budget.
+    uint64_t memoryUnderBudget = GetBudgetLeft(gResidencyManager.Get(), bufferMemorySegment);
+    while (gResourceAllocator->GetInfo().UsedMemoryUsage + kBufferMemorySize < memoryUnderBudget) {
```
[the remark this answers is not in the document]
Comment (bbernhar) [reply src/fuzzers/D3D12ResidencyManagerFuzzer.cpp:90]: I will go over the budget once LLVMFuzzerTestOneInput is called. 

How far "beyond" the budget should I go? 

If I go too far, LLVMFuzzerInitialize could fail and the fuzzer won't run. =(

[the remark this answers is not in the document]
Comment (bjjones) [reply src/fuzzers/D3D12ResidencyManagerFuzzer.cpp:90]: Isn't the point of fuzzing to be able to fail gracefully in these unrealistic scenarios? There's nothing stopping a bad actor from calling CreateResource until it crashes - so I don't think the fuzzer should be limited in that way either.

> How far "beyond" the budget should I go?

It seems like the most correct operation would be to keep allocating until failure. A passing test is one that failed cleanly as opposed to one that failed unexpectedly.

Your call though - I'm not a fuzzing expert and this is just my understanding of how it is supposed to work.

[the remark this answers is not in the document]
Comment (bjjones) [reply src/fuzzers/D3D12ResidencyManagerFuzzer.cpp:90]: Note this test is different from the fuzzing issues I've encountered in Dawn - where they usually are just providing garbage inputs into the entry points. This test is waaay deeper into testing the functionality than the Dawn fuzzing I've seen. 

...And maybe the answer lies in that? Are you sure this needs to be fuzzed and isn't more appropriate as an e2e test specifically for catching odd sized over commitment?

[the remark this answers is not in the document]
Comment (bbernhar) [reply src/fuzzers/D3D12ResidencyManagerFuzzer.cpp:90]: Fuzzer is meant to catch ungraceful failures (crashes, timeouts), not OOM. Already test CreateResource going OOM, fuzzed separately, as residency is optional. Like Dawn, this fuzzer also provides garbage inputs but I can't cover the residency LRUs without reaching the budget first (missing code to cover). Only once the fuzzer crashes, you can create e2e from the offending input (but not the other way around).

Does that make more sense?

[the remark this answers is not in the document]
Comment (bjjones) [reply src/fuzzers/D3D12ResidencyManagerFuzzer.cpp:90]: > Only once the fuzzer crashes, you can create e2e from the offending input (but not the other way around).

I was thinking there are so few corner cases in the residency manager that it'd make more sense to just write tests for them. But I suppose it isn't going to hurt to set it up for the fuzzer.

**Review by bjjones — APPROVED** [review]
LGTM
